### PR TITLE
[ADD] Payment_authorize: add refund for Authorize.

### PR DESCRIPTION
**Pull request by hote-odoo**
[ADD] payment_authorize: add refund for Authorize.
    
    Users can now ask for a refund from Odoo for their transactions done
    through Authorize.net. A refund will be triggered from Odoo when
    necessary.
    Only full refunds are possible.
    Note that unsettled transactions will be voided as they cannot be
    refunded.

Task - 2678757

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/payment_authorize/models/payment_transaction.py [hunk 1]
```diff
@@ -95,18 +95,37 @@ def _send_refund_request(self, amount_to_refund=None, create_refund_transaction=
...
         authorize_API = AuthorizeAPI(self.acquirer_id)
-        rounded_amount = round(self.amount, self.currency_id.decimal_places)
-        res_content = authorize_API.refund(self.acquirer_reference, rounded_amount)
+        tx_details = authorize_API.get_transaction_details(self.acquirer_reference)
+
+        if tx_details and tx_details.get('err_code'):
+            res_content = {
+                'x_response_code': AuthorizeAPI.AUTH_ERROR_STATUS,
+                'x_response_reason_text': tx_details.get('err_msg')
+            }
```
Comment (Demesmaeker) [inline addons/payment_authorize/models/payment_transaction.py:107]: Does this always means it's a refund for a validation operation? Can't we just check the `self.operation` instead, if so?
You can add a comment to explain this.

File: addons/payment_authorize/models/payment_transaction.py [hunk 2]
```diff
@@ -95,18 +95,37 @@ def _send_refund_request(self, amount_to_refund=None, create_refund_transaction=
...
+            )
+        elif tx_details.get('transaction', {}).get('transactionStatus') in ['capturedPendingSettlement', 'authorizedPendingCapture']:
+            refund_tx = super()._send_refund_request(
+                amount_to_refund=self.amount, create_refund_transaction=True
+            )
+            res_content = authorize_API.void(self.acquirer_reference)
+        else:
+            refund_tx = super()._send_refund_request(
+                amount_to_refund=amount_to_refund, create_refund_transaction=True
+            )
```
Comment (Demesmaeker) [inline addons/payment_authorize/models/payment_transaction.py:120]: Same here, you can add a comment with something like:
"Pending tx can't be refund, they must be void."

File: addons/payment_authorize/models/payment_transaction.py [hunk 3]
```diff
@@ -95,18 +95,37 @@ def _send_refund_request(self, amount_to_refund=None, create_refund_transaction=
...
+            # We continue to use the full amount below
+            rounded_amount = round(amount_to_refund, self.currency_id.decimal_places)
+            res_content = authorize_API.refund(self.acquirer_reference, rounded_amount)
+
         _logger.info("refund request response:\n%s", pprint.pformat(res_content))
-        # As the API has no redirection flow, we always know the reference of the transaction.
-        # Still, we prefer to simulate the matching of the transaction by crafting dummy feedback
-        # data in order to go through the centralized `_handle_feedback_data` method.
-        feedback_data = {'reference': self.reference, 'response': res_content}
+        feedback_data = {'reference': refund_tx.reference, 'response': res_content}
```
Comment (Demesmaeker) [inline addons/payment_authorize/models/payment_transaction.py:128]: In a validation case, the refund_tx will be self.env['payment.transaction'], not the validation tx reference ;)

```suggestion
        feedback_data = {'reference': refund_tx.reference or self.reference, 'response': res_content}
```

File: addons/payment_authorize/models/payment_transaction.py [hunk 4]
```diff
@@ -95,18 +95,37 @@ def _send_refund_request(self, amount_to_refund=None, create_refund_transaction=
                 create_refund_transaction=create_refund_transaction,
             )
 
-        refund_tx = super()._send_refund_request(
-            amount_to_refund=amount_to_refund, create_refund_transaction=False
-        )
+        # Only full refund is possible.
```
Comment (Demesmaeker) [inline addons/payment_authorize/models/payment_transaction.py:98]: I feel like this is useless, since the `support_refund` is set as `full_only`

**Review by Demesmaeker — COMMENTED** [review]
Good job here =]
Don't forget to merge your 3 commits into one, you can also add the reference of your task with `task-2678757` at the end of the description/commit. (Edit: It's easier to find the related task for someone who is browsing the commits history for instance) Also, you don't need to finish the title of your commits with `.` :)

File: addons/payment_authorize/models/payment_transaction.py (hunk 4 above)
Comment (Demesmaeker) [reply addons/payment_authorize/models/payment_transaction.py:98]: I mean the comment (didn't expect the review to show it like this >.<)

File: addons/payment/controllers/portal.py [hunk 5]
```diff
@@ -159,9 +159,14 @@ def payment_method(self, **kwargs):
...
-        tokens = set(partner.payment_token_ids).union(
-            partner.commercial_partner_id.sudo().payment_token_ids
-        )  # Show all partner's tokens, regardless of which acquirer is available
+
+        # Get all partner's tokens, regardless of which acquirers are available and then
+        # filter tokens for disabled acquirers and remove duplicates.
+        tokens = set((
+                             partner.payment_token_ids +
+                             partner.commercial_partner_id.sudo().payment_token_ids
+                     ).filtered(lambda record: record.acquirer_id.state != 'disabled'))
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/payment/controllers/portal.py:168]: I believe this is an old version of a commit from another task.  I don't think this is related to the refund :thinking:

[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/controllers/portal.py:168]: It's not, it's in two others commit actually, I thought it was a bugfix he was asked to make at the same time?

[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/controllers/portal.py:168]: task-2679695

So yes, nothing to do here, it's from another branch ^^ You can take this off

File: addons/payment_authorize/models/payment_transaction.py (hunk 3 above)
Comment (hote-odoo) [reply addons/payment_authorize/models/payment_transaction.py]: done :)

File: addons/payment_authorize/models/payment_transaction.py (hunk 4 above)
Comment (hote-odoo) [reply addons/payment_authorize/models/payment_transaction.py]: done :)

File: addons/payment_authorize/models/payment_transaction.py (hunk 2 above)
Comment (hote-odoo) [reply addons/payment_authorize/models/payment_transaction.py]: done :)

File: addons/payment_authorize/models/payment_transaction.py (hunk 1 above)
Comment (hote-odoo) [reply addons/payment_authorize/models/payment_transaction.py]: done :)

File: addons/payment_authorize/models/payment_transaction.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/payment_authorize/models/payment_transaction.py]: done :)

File: addons/payment/controllers/portal.py [hunk 6]
```diff
@@ -159,9 +159,18 @@ def payment_method(self, **kwargs):
...
+        # filter tokens for disabled acquirers and remove duplicates.
+        # We use 'sudo' here to access tokens_ids regardless of the user logged in at the moment.
+        partner_tokens = (partner.payment_token_ids
+                          + partner.commercial_partner_id.sudo().payment_token_ids)
+        # There is no need of a 'sudoed' environments for the tokens, as the information is
+        # already contained in the partner_tokens and we are are only filtering data that
+        # we already got.
+        tokens = set(partner_tokens
+                     .filtered(lambda record: record.acquirer_id.state != 'disabled'))
+
```
Comment (Feyensv) [inline addons/payment/controllers/portal.py:173]: Still not the target of the current task AFAIK.
This diff should be removed from current PR (you already made a better version in #79253, and it'll be merged before).

Comment (hote-odoo) [reply addons/payment/controllers/portal.py]: Thanks for the explanation, I did not understand before, hopefully it is as before I made  the changes.

Comment (Feyensv) [reply addons/payment/controllers/portal.py]: :+1:

File: addons/payment_authorize/models/payment_transaction.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment_authorize/models/payment_transaction.py]: When you finish with a suggestion/comment/... you can mark it as "resolved" ;)
But you can let it unresolved if you asked another question/are waiting for an answer/didn't change but justified why... this way, the reviewer knows what still needs it's attention =]

File: addons/payment_authorize/models/payment_transaction.py [hunk 7]
```diff
@@ -95,18 +95,36 @@ def _send_refund_request(self, amount_to_refund=None, create_refund_transaction=
...
+              in ['capturedPendingSettlement', 'authorizedPendingCapture']):
+            # For Authorize captured transactions are not automotically debited.
+            # Transactions with status in 'captured, pending settlement' and 'authorised,
+            # pending capture' are voided if a refund is asked, as the credit was not moved
+            # yet.
+            res_content = authorize_API.void(self.acquirer_reference)
+        else:
+            # continue with the usual refund
+            # We continue to use the full amount below
+            rounded_amount = round(amount_to_refund, self.currency_id.decimal_places)
```
Comment (Demesmaeker) [inline addons/payment_authorize/models/payment_transaction.py:123]: To be consistent, you may want to choose between `amount_to_refund` (which will be the full amount due to the fact it's `full_only`), or use `self.amount` as in line 103. 
Also, is the comment needed here? If you feel it's, then don't mind me. :)

**Review by Demesmaeker — CHANGES_REQUESTED** [review]

File: addons/payment_authorize/data/payment_acquirer_data.xml [hunk 8]
```diff
@@ -6,7 +6,7 @@
         <field name="inline_form_view_id" ref="inline_form"/>
         <field name="support_authorization">True</field>
         <field name="support_fees_computation">False</field>
-        <field name="support_refund"></field>
+        <field name="support_refund">full_only</field>
```
Comment (Demesmaeker) [inline addons/payment_authorize/data/payment_acquirer_data.xml:9]: This'll need an Upgrade PR =]

Comment (hote-odoo) [reply addons/payment_authorize/data/payment_acquirer_data.xml:9]: I have the upgrade PR ready, I will wait until the PR here is almost done and then push.

File: addons/payment_authorize/models/authorize_request.py [hunk 9]
```diff
@@ -259,7 +259,7 @@ def _prepare_tx_data(self, token=None, opaque_data=False):
                 }
             }
 
-    def _get_transaction_details(self, transaction_id):
+    def get_transaction_details(self, transaction_id):
```
Comment (Demesmaeker) [inline addons/payment_authorize/models/authorize_request.py:262]: Why did you choose to change this?

File: addons/payment_authorize/models/payment_transaction.py [hunk 10]
```diff
@@ -96,18 +96,42 @@ def _send_refund_request(self, amount_to_refund=None, create_refund_transaction=
...
-        self._handle_notification_data('authorize', {'response': res_content})
+        if tx_details and tx_details.get('err_code'):
+            # In the case of an error a 'res_content' must be passed back to the
+            # '_handle_feedback_data' anyways.
+            res_content = {
+                'x_response_code': AuthorizeAPI.AUTH_ERROR_STATUS,
+                'x_response_reason_text': tx_details.get('err_msg')
+            }
+        elif (tx_details.get('transaction', {}).get('transactionStatus')
+              in ['capturedPendingSettlement', 'authorizedPendingCapture']):
```
Comment (Demesmaeker) [inline addons/payment_authorize/models/payment_transaction.py:118]:
```suggestion
        elif (
            tx_details.get('transaction', {}).get('transactionStatus')
            in ['capturedPendingSettlement', 'authorizedPendingCapture']
        ):
```
or
```suggestion
        tx_status = tx_details.get('transaction', {}).get('transactionStatus')
        ...
        elif (tx_status in ['capturedPendingSettlement', 'authorizedPendingCapture']):
```

File: addons/payment_authorize/models/payment_transaction.py [hunk 11]
```diff
@@ -96,18 +96,42 @@ def _send_refund_request(self, amount_to_refund=None, create_refund_transaction=
...
 
-        authorize_API = AuthorizeAPI(self.acquirer_id)
-        rounded_amount = round(self.amount, self.currency_id.decimal_places)
-        res_content = authorize_API.refund(self.acquirer_reference, rounded_amount)
-        _logger.info(
-            "refund request response for transaction with reference %s:\n%s",
-            self.reference, pprint.pformat(res_content)
-        )
-        self._handle_notification_data('authorize', {'response': res_content})
+        if tx_details and tx_details.get('err_code'):
```
Comment (Demesmaeker) [inline addons/payment_authorize/models/payment_transaction.py:110]: You check here if the `tx_details` exists, but not in the `elif`. ;)

File: addons/payment_authorize/models/payment_transaction.py [hunk 12]
```diff
@@ -96,18 +96,42 @@ def _send_refund_request(self, amount_to_refund=None, create_refund_transaction=
...
+            # yet.
+            res_content = authorize_API.void(self.acquirer_reference)
+        else:
+            # If the transaction has been settled on the Authorize side then an actual
+            # refund can be made.
+            rounded_amount = round(amount_to_refund, self.currency_id.decimal_places)
+            res_content = authorize_API.refund(self.acquirer_reference, rounded_amount)
+
+        _logger.info("refund request response for transaction with reference %s:\n%s",
+                     self.acquirer_reference, pprint.pformat(res_content))
```
Comment (Demesmaeker) [inline addons/payment_authorize/models/payment_transaction.py:131]:
```suggestion
        _logger.info(
            "refund request response for transaction with reference %s:\n%s",
            self.acquirer_reference, 
            pprint.pformat(res_content)
        )
```

**Review by Demesmaeker — COMMENTED** [review]
=P (I know, it was just a rebase... :innocent:  )

File: addons/payment_authorize/models/authorize_request.py (hunk 9 above)
Comment (hote-odoo) [reply addons/payment_authorize/models/authorize_request.py:262]: I using the function outside of the module. I think in that case is better to "make" the method public.

File: addons/payment_authorize/models/payment_transaction.py (hunk 10 above)
Comment (hote-odoo) [reply addons/payment_authorize/models/payment_transaction.py]: done, used your first suggestion :)

File: addons/payment_authorize/models/payment_transaction.py (hunk 12 above)
Comment (hote-odoo) [reply addons/payment_authorize/models/payment_transaction.py]: done :)

File: addons/payment_authorize/models/payment_transaction.py (hunk 11 above)
Comment (hote-odoo) [reply addons/payment_authorize/models/payment_transaction.py:110]: You're right!, I fixed it :)

**Pull request closed without merging**